lib/pbench/test/unit/server/test_schema.py

```diff
@@ -0,0 +1,153 @@
+import dateutil
+import pytest
+from typing import Callable
+
+from pbench.server.api.auth import Auth, UnknownUser
+from pbench.server.api.resources.query_apis import (
+    ParamType,
+    ConversionError,
+    InvalidRequestPayload,
+    MissingParameters,
+    Parameter,
+    Schema,
+)
+
+
+class TestParamType:
+    """
+    Tests on the ParamType enum
+    """
+
+    def test_enum(self):
+        assert (
+            len(ParamType.__members__) == 4
+        ), "Number of ParamType ENUM values has changed; confirm test coverage!"
+        for n, t in ParamType.__members__.items():
+            assert str(t) == t.friendly.upper()
+            assert isinstance(t.convert, Callable)
+
+    @pytest.mark.parametrize(
+        "test",
+        (
+            (ParamType.STRING, "x", "x"),
+            (ParamType.JSON, {"key": "value"}, {"key": "value"}),
+            (ParamType.DATE, "2021-06-29", dateutil.parser.parse("2021-06-29")),
+            (ParamType.USER, "drb", "drb"),
+        ),
+    )
+    def test_successful_conversions(self, test, monkeypatch):
+        def ok(user: str) -> str:
+            return user
+
+        monkeypatch.setattr(Auth, "validate_user", ok)
+
+        ptype, value, expected = test
+        result = ptype.convert(value)
+        assert result == expected
+
+    @pytest.mark.parametrize(
+        "test",
+        (
+            (ParamType.STRING, {"not": "string"}),
+            (ParamType.JSON, "not_json"),
+            (ParamType.DATE, "2021-06-45"),
+            (ParamType.USER, "drb"),
+        ),
+    )
+    def test_failed_conversions(self, test, monkeypatch):
+        def not_ok(user: str) -> str:
+            raise UnknownUser()
+
+        monkeypatch.setattr(Auth, "validate_user", not_ok)
+
+        ptype, value = test
+        with pytest.raises(ConversionError) as exc:
+            ptype.convert(value)
+        assert str(exc).find(str(value))
+
+
+class TestParameter:
+    """
+    Tests on the Parameter class
+    """
+
+    def test_constructor(self):
+        x = Parameter("test", ParamType.STRING)
+        assert not x.required
+        assert x.name == "test"
+        assert x.type is ParamType.STRING
+
+        y = Parameter("foo", ParamType.JSON, required=True)
+        assert y.required
+        assert y.name == "foo"
+        assert y.type is ParamType.JSON
```

Comment:
No love for `DATE` or `USER`?  (Seems like an excellent spot for one of your `pytest.mark.parametrize` gizmos?  BTW, I have to say, that's a d\*mned annoying spelling, that is.)

Reply:
I already checked that all of the expected ENUM types are defined. (Well, at least that the count is right.) This test doesn't care about the specific types, since it's not doing any conversion; the intent was just to check that the observed object matches the constructor parameters, so it should apply to any type. I only have two cases here because of the optional `required` parameter.

Reply:
Ah...I see -- you're testing "required" vs. "not-required".  In that case, I'd recommend using the same `ParamType` in each scenario.  (But, don't hold the merge for that.)

Reply:
Technically I'm just testing the constructor, as I tried to imply by the test case name. Yeah, it wasn't necessary to use different parameter types, but I'll always take testing variety over monotony when there's a choice. I should, however, have written a docstring for each test case; and while I do often remember to practice that discipline, I offer these tests as humble evidence that I'm not very good at it yet ... ;-)